File: src/main/java/fi/vm/sade/lokalisointi/configuration/ServletContainerConfiguration.java
package fi.vm.sade.lokalisointi.configuration;

import org.apache.catalina.connector.Connector;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
import org.springframework.boot.tomcat.servlet.TomcatServletWebServerFactory;
import org.springframework.boot.web.server.WebServerFactoryCustomizer;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.Profile;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.nio.file.Files;

@Configuration
@Profile("!dev")
@ConditionalOnProperty("lokalisointi.uses-ssl-proxy")
public class ServletContainerConfiguration
    implements WebServerFactoryCustomizer<TomcatServletWebServerFactory> {
  private static final Logger LOG = LoggerFactory.getLogger(ServletContainerConfiguration.class);

  /**
   * Konfiguraatio kun palvelua ajetaan HTTPS proxyn läpi. Käytännössä tämä muuttaa {@link
   * javax.servlet.ServletRequest#getScheme()} palauttamaan `https` jolloin palvelun kautta luodut
   * urlit muodostuvat oikein.
   *
   * <p>Aktivointi: `lokalistointi.uses-ssl-proxy` arvoon `true`.
   */
  @Override
  public void customize(final TomcatServletWebServerFactory server) {
    server.addConnectorCustomizers(
        (Connector connector) -> {
          connector.setScheme("https");
          connector.setSecure(true);
        });
  }
}

File: src/main/java/fi/vm/sade/lokalisointi/configuration/ServletContainerConfiguration.kt
package fi.vm.sade.lokalisointi.configuration

import org.apache.catalina.connector.Connector
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty
import org.springframework.boot.web.embedded.tomcat.TomcatConnectorCustomizer
import org.springframework.boot.web.embedded.tomcat.TomcatServletWebServerFactory
import org.springframework.boot.web.server.WebServerFactory
import org.springframework.boot.web.server.WebServerFactoryCustomizer
import org.springframework.boot.web.servlet.server.ConfigurableServletWebServerFactory
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration

@Configuration
class ServletContainerConfiguration {
    /**
     * Konfiguraatio kun palvelua ajetaan HTTPS proxyn läpi. Käytännössä tämä muuttaa [ ][javax.servlet.ServletRequest.getScheme] palauttamaan `https` jolloin palvelun kautta luodut
     * urlit muodostuvat oikein.
     *
     *
     * Aktivointi: `lokalistointi.uses-ssl-proxy` arvoon `true`.
     *
     * @return EmbeddedServletContainerCustomizer jonka Spring automaattisesti tunnistaa ja lisää
     * servlet containerin konfigurointiin
     */
    @Bean
    @ConditionalOnProperty("lokalisointi.uses-ssl-proxy")
    fun sslProxyCustomizer(): WebServerFactoryCustomizer<*> {
        return WebServerFactoryCustomizer { container: WebServerFactory? ->
            if (container is ConfigurableServletWebServerFactory) {
                val tomcat = container as TomcatServletWebServerFactory
                tomcat.addConnectorCustomizers(
                    TomcatConnectorCustomizer { connector: Connector ->
                        connector.scheme = "https"
                        connector.secure = true
                    })
            }
        }
    }
}
